dm/master/simulator.go

```diff
@@ -0,0 +1,584 @@
+		relativeEventFilterMap := make(map[string]*bf.BinlogEventRule, len(mysqlInstance.FilterRules))
+		for _, eventFilterName := range mysqlInstance.FilterRules {
+			relativeEventFilterMap[eventFilterName] = cfg.Filters[eventFilterName]
+		}
+
+		filterName, action, err := filterSQL(sql, relativeEventFilterMap, stCfg.CaseSensitive)
+		if err != nil {
+			return &pb.SimulationResponse{
+				Result: false,
+				Msg:    errors.ErrorStack(err),
+			}, nil
+		}
+
+		simulationResult := &pb.SimulationResult{
+			SourceID:   stCfg.SourceID,
+			SourceAddr: getSourceAddr(stCfg.From),
+		}
+		if action == bf.Ignore {
+			simulationResult.IgnoreTableMap = map[string]*pb.TableList{sql: {Reasons: []string{filterName}}}
+		} else {
+			simulationResult.DoTableMap = map[string]*pb.TableList{sql: {Reasons: []string{filterName}}}
+		}
+		simulationResults = append(simulationResults, simulationResult)
+		break
+	}
+	resp.SimulationResults = simulationResults
+	return resp, nil
+}
+
+func getSourceAddr(dbCfg config.DBConfig) string {
+	return fmt.Sprintf("%s:%d", dbCfg.Host, dbCfg.Port)
+}
+
+func getMySQLInstanceConfigThroughSourceID(mysqlInstances []*config.MySQLInstance, sourceID string) *config.MySQLInstance {
+	var mysqlInstance *config.MySQLInstance
+	for _, mysqlInst := range mysqlInstances {
+		if mysqlInst.SourceID == sourceID {
+			mysqlInstance = mysqlInst
+			break
+		}
+	}
+	return mysqlInstance
+}
+
+func checkSingleBWFilter(schema, table string, bw *filter.Filter) bool {
+	return len(bw.ApplyOn([]*filter.Table{{Schema: schema, Name: table}})) == 0
+}
+
+func checkSingleBWRules(schema, table string, caseSensitive bool, rules *filter.Rules) (bool, error) {
+	bwFilter, err := filter.New(caseSensitive, rules)
+	if err != nil {
+		return false, err
+	}
+	return checkSingleBWFilter(schema, table, bwFilter), nil
+}
+
+// getRouteLevel gets route result of whether schema table can be routed by r
+// 0: won't be routed; 1: match schema rule; 2: match table rule
+func getRouteLevel(r *router.Table, caseSensitive bool, schema, table string) (int, error) {
+	schemaL, tableL := schema, table
+	if !caseSensitive {
+		schemaL, tableL = strings.ToLower(schema), strings.ToLower(table)
+	}
+
+	rules := r.Match(schemaL, tableL)
+	var (
+		schemaRules = make([]*router.TableRule, 0, len(rules))
+		tableRules  = make([]*router.TableRule, 0, len(rules))
+	)
+	// classify rules into schema level rules and table level
+	// table level rules have highest priority
+	for i := range rules {
+		rule, ok := rules[i].(*router.TableRule)
+		if !ok {
+			return 0, errors.NotValidf("table route rule %+v", rules[i])
+		}
+
+		if len(rule.TablePattern) == 0 {
+			schemaRules = append(schemaRules, rule)
+		} else {
+			tableRules = append(tableRules, rule)
+		}
+	}
+
+	if len(table) == 0 || len(tableRules) == 0 {
+		if len(schemaRules) > 1 {
+			return 0, errors.NotSupportedf("route %s/%s to rule set(%d)", schema, table, len(schemaRules))
```

Comment:
also put matched rules in the error message, which will help user to find conflict rules directly.